### PR TITLE
Generate protobuf schema

### PR DESCRIPTION
**Pull request by alexanderst**

### CODE REVIEW

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java [hunk 1]
```diff
@@ -112,12 +125,117 @@ public void testEmbeddedData() {
...
-    @Ignore
     @Test
     public void testSerde() {
         Data someData = JData.singleton(KEY, Long.valueOf(200));
-        someData.save(new File("temp"));
-        Data restoredData = JData.fromFile(new File("temp"));
+        try {
+            someData.save(new File("temp"));
+        } catch (IOException e) {
+            log.error("IOException in DataTest.testSerde()", e);
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java:134]: 2 problems here
(a) Don't use `new File("...")` in tests - this is something I flagged in the previous PR here: https://github.com/KonduitAI/konduit-serving/pull/297#discussion_r417772653
(b) Don't simply log exceptions, this is very bad... an exception can be thrown and the test can still pass.
In fact, we don't need try try/catch blocks at all - just use `public void testSerde() throws Exception` and let any exceptions propagate out and fail the test if they occur

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java [hunk 2]
```diff
@@ -112,12 +125,117 @@ public void testEmbeddedData() {
...
+            restoredData = JData.fromFile(new File("temp"));
+        } catch (IOException e) {
+            log.error("IOException in DataTest.testSerde()", e);
+        }
+        assertEquals(someData.get(KEY), restoredData.get(KEY));
+    }
+
+    @Test
+    public void testConvertToBytes() {
+        JData someData = (JData) JData.singleton(KEY, Long.valueOf(200));
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java:147]: `Data someData = Data.singleton(KEY, 200L);`
Thought let's also handle the Integer -> Long internally too for this method

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java [hunk 3]
```diff
@@ -112,12 +125,117 @@ public void testEmbeddedData() {
...
+                "  \"mapItems\": {\n" +
+                "    \"stringData\": {\n" +
+                "      \"iValue\": \"200\",\n" +
+                "      \"type\": \"INT64\"\n" +
+                "    }\n" +
+                "  }\n" +
+                "}";
+
+        assertEquals(expected, jsonStr);
+    }
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java:166]: Test can go, JSON is implemented/tested elsewhere

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java [hunk 4]
```diff
@@ -141,4 +259,44 @@ public void testBuilderVSFactory() {
...
+    public void testStreamsSerde() {
+        JData someData = (JData)Data.singleton(KEY, "testString");
+        Data restored = empty();
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try {
+            someData.write(baos);
+            restored = Data.fromStream(new ByteArrayInputStream(baos.toByteArray()));
+        } catch (IOException e) {
+            log.error("testStreamsSerde failed", e);
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java:298]: Bunch more try/catch + exception issues as mentioned above.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 5]
```diff
@@ -83,7 +95,10 @@ public ValueType type(String key) {
 
     @Override
     public ValueType listType(String key) {
-        return null;
+        Value data = dataMap.get(key);
+        if (data == null || !(data instanceof ListValue))
+            throw new ValueNotFoundException(String.format("Value not found for key %s", key));
```
[the remark this answers is not in the document]
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java:82]: It makes sense after changing string literal to constant string. It allows to move error text to properties file and localize.

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java (hunk 2 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java]: Shouldn't we add int32 and float data types? We are dealing with Object ref and upcast is not possible.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 6]
```diff
@@ -43,7 +47,15 @@ public int size() {
...
-        throw new UnsupportedOperationException();
+        StringBuilder jsonString = new StringBuilder();
+        try {
+            generated.Data.DataMap pbDataMap = javaObjectToPbMessage();
+            jsonString.append(JsonFormat.printer().print(pbDataMap));
+
+        } catch (InvalidProtocolBufferException e) {
+            log.error("Failed toJson conversion",e);
+        }
+        return jsonString.toString();
```
[the remark this answers is not in the document]
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 7]
```diff
@@ -208,22 +250,175 @@ public void setMetaData(Data data) {
     }
 
     @Override
-    public void save(File toFile) {
+    public void save(File toFile) throws IOException {
+        write(new FileOutputStream(toFile));
```
[the remark this answers is not in the document]
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java]: Removed method.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 8]
```diff
@@ -208,22 +250,175 @@ public void setMetaData(Data data) {
...
+        generated.Data.DataMap.Builder builder = generated.Data.DataMap.newBuilder().mergeFrom(new FileInputStream(fromFile));
+        generated.Data.DataMap dataMap = builder.build();
+        return pbMapToJavaData(dataMap);
+    }
+
+    public static Data fromStream(InputStream stream) throws IOException {
+        generated.Data.DataMap.Builder builder = generated.Data.DataMap.newBuilder().mergeFrom(stream);
+        generated.Data.DataMap dataMap = builder.build();
+        return pbMapToJavaData(dataMap);
+    }
```
[the remark this answers is not in the document]
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java]: Removed method.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/wrappers/ListValue.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package ai.konduit.serving.pipeline.impl.data.wrappers;
+
+import ai.konduit.serving.pipeline.impl.data.Value;
+import ai.konduit.serving.pipeline.impl.data.ValueType;
+import lombok.AllArgsConstructor;
+
+import java.util.List;
+
+@AllArgsConstructor
+public class ListValue<T> implements Value<List<T>> {
```
[the remark this answers is not in the document]
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/wrappers/ListValue.java:10]: Changed, but ListValue.set() method is not uniform.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 10]
```diff
@@ -208,22 +250,175 @@ public void setMetaData(Data data) {
...
+        try {
+            generated.Data.DataMap.Builder builder = generated.Data.DataMap.newBuilder();
+            JsonFormat.parser().merge(jsonString, builder);
+            generated.Data.DataMap dataMap = builder.build();
+            retVal = pbMapToJavaData(dataMap);
+        } catch (InvalidProtocolBufferException e) {
+            log.error("Failed toJson conversion",e);
+        }
+        return retVal;
     }
```
[the remark this answers is not in the document]
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 11]
```diff
@@ -192,6 +214,26 @@ public void put(String key, Data data) {
...
+
+    @Override
+    public void putListBoolean(String key, List<Boolean> data) {
+        dataMap.put(key, new ListValue(data));
+    }
+
+    @Override
+    public void putListDouble(String key, List<Double> data) {
+        dataMap.put(key, new ListValue(data));
+    }
```
[the remark this answers is not in the document]
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java:228]: Fixed.

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java (hunk 1 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java]: Fixed.

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java (hunk 3 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java:144]: Removed test.

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java (hunk 4 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java]: Fixed.

File: konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java (hunk 2 above)
Comment (AlexDBlack) [reply konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java]: Let's do Float -> Double, and Integer -> Long
upcast is definitely possible, in 2 ways:
Use Number.longValue() and similar methods (all boxed primites are Number objects - Double, Float, Integer, Long, etc)
Cast to primitive:
double d = (Float)obj;

Comment (alexanderst) [reply konduit-serving-pipeline/src/test/java/ai/konduit/serving/pipeline/impl/data/DataTest.java]: Fixed with upcast.

File: konduit-serving-io/konduit-serving-camera/src/main/java/ai/konduit/serving/camera/step/capture/FrameCapturePipelineStep.java [hunk 12]
```diff
@@ -1,4 +1,4 @@
-/*
+n/*
```
Comment (AlexDBlack) [inline konduit-serving-io/konduit-serving-camera/src/main/java/ai/konduit/serving/camera/step/capture/FrameCapturePipelineStep.java:1]: Remove

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/protobuf/data.proto [hunk 13]
```diff
@@ -0,0 +1,62 @@
+
+syntax = "proto3";
+
+package generated;
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/protobuf/data.proto:4]:
```
package = ai.konduit.serving;
option java_package = "ai.konduit.serving.pipeline.impl.data.protobuf";
option java_outer_classname = "DataProtoMessage"
```

File: konduit-serving-api/src/test/java/ai/konduit/serving/data/GeneratedClassesTest.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
+package ai.konduit.serving.data;
```
Comment (AlexDBlack) [inline konduit-serving-api/src/test/java/ai/konduit/serving/data/GeneratedClassesTest.java:1]: Move all classes/tests added in this PR from konduit-serving-api to konduit-serving-pipeline - everything should be in that module, no additions to konduit-serving-api

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java [hunk 15]
```diff
@@ -250,6 +307,36 @@ else if (data instanceof Data) {
         return instance;
     }
 
+    public static Data singletonList(@NonNull String key, @NonNull List<?> data,
+                                     @NonNull ValueType valueType) {
+        Data instance = new JData();
+        if (valueType.equals(ValueType.STRING)) {
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java:313]: ValueType is an enum - use `if(valueType == ValueType.STRING)` etc here instead.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    public ProtoData(@NonNull File file) throws IOException {
+        this(fromFile(file));
+    }
+
+    public ProtoData(@NonNull Data data) {
+        if(data instanceof JData){
+            getDataMap().putAll(((JData)data).getDataMap());
+        }
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java:46]: for the "else" case here let's throw an UnsupportedOperationException
Won't do anything now - but if we add a new Data implementation and forget about this - we'll know right away.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Override
+    public ProtoData toProtoData() {
+        return this;
+    }
+
+    @Override
+    public void save(File toFile) throws IOException {
+        write(new FileOutputStream(toFile));
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java:56]: (a) buffer, and (b) make sure to close your streams.
Anytime you are dealing with files, you should _always_ buffer and close streams to ensure file handle is released.
```
try(OutputStream os = new BufferedOutputStream(new FileOutputStream(toFile))){
```

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+                ProtobufUtils.convertJavaToProtobufData(getDataMap());
+
+        generated.Data.DataMap pbDataMap = generated.Data.DataMap.newBuilder().
+                putAllMapItems(newItemsMap.getMapItemsMap()).
+                build();
+        pbDataMap.writeTo(toStream);
+    }
+
+    public static Data fromFile(File fromFile) throws IOException {
+        generated.Data.DataMap.Builder builder = generated.Data.DataMap.newBuilder().mergeFrom(new FileInputStream(fromFile));
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java:72]: can replace with:
```
try(InputStream is = new FileInputStream(fromFile)){
    return fromStream(is);
}
```
Normally we'd use a BufferedInputStream here too (critical for performance) but it looks like mergeFrom uses CodedInputStream internally which is buffered. Maybe add a comment that "mergeFrom is buffered internally".

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+        return ProtobufUtils.convertProtobufToData(dataMap);
+    }
+
+
+    public byte[] asBytes() {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try {
+            write(baos);
+        } catch (IOException e) {
+            log.error("Failed write to ByteArrayOutputStream", e);
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java:89]: Don't just log the exception. Throw a DataLoadingException instead (recently added on master)

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+        return baos.toByteArray();
+    }
+
+    public static Data fromBytes(byte[] input) {
+        Data retVal = empty();
+        generated.Data.DataMap.Builder builder = null;
+        try {
+            builder = generated.Data.DataMap.newBuilder().mergeFrom(input);
+        } catch (InvalidProtocolBufferException e) {
+            log.error("Error converting bytes array to data",e);
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java:100]: As above - exception

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java [hunk 21]
```diff
@@ -0,0 +1,125 @@
+package ai.konduit.serving.pipeline.impl.data.helpers;
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:1]: Copyright header

Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/helpers/ProtobufUtils.java:18]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java (hunk 20 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java (hunk 19 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java (hunk 18 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java (hunk 17 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java (hunk 16 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/ProtoData.java:47]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java (hunk 15 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/impl/data/JData.java]: Fixed.

File: konduit-serving-api/src/test/java/ai/konduit/serving/data/GeneratedClassesTest.java (hunk 14 above)
Comment (alexanderst) [reply konduit-serving-api/src/test/java/ai/konduit/serving/data/GeneratedClassesTest.java]: Moved. This file is to be removed - just used it to verify successful code generation. Looks like it doesn't make sense to test this API, covered by higher level tests anyway.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/protobuf/data.proto (hunk 13 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/protobuf/data.proto]: Fixed.

File: konduit-serving-io/konduit-serving-camera/src/main/java/ai/konduit/serving/camera/step/capture/FrameCapturePipelineStep.java (hunk 12 above)
Comment (alexanderst) [reply konduit-serving-io/konduit-serving-camera/src/main/java/ai/konduit/serving/camera/step/capture/FrameCapturePipelineStep.java]: Fixed.

**Pull request merged**